### ISSUE TITLE
Garden dev behaviour differs from deploy

### ISSUE THREAD

**Issue by drubin** [garden-io/garden#388]
The calculation of tasks differs between `dev` and `deploy` in unexpected ways. This was partially fixed here https://github.com/garden-io/garden/pull/370 

My understanding is that `garden deploy -w` behaves the same as `garden dev` but with out tests? 

---

This workflow is using the `gatsby-hot-reload` example project

Each one of these hot reloading + rebuilding were done because I ran 

`garden get status` (See #387)

```sh
garden dev --hot-reload=website

Good morning! Let's get your environment wired up...

✔ website                   → Building website:2c12317af0... → Done (took 36.2 sec)
✔ website                   → Deploying → Ready
✔ website                   → Hot reloading → Done (took 316 ms)
✔ website                   → Building website:2c12317af0... → Done (took 0.7 sec)
✔ website                   → Hot reloading → Done (took 144 ms)
✔ website                   → Building website:2c12317af0... → Done (took 0.4 sec)
✔ website                   → Hot reloading → Done (took 132 ms)
✔ website                   → Building website:2c12317af0... → Done (took 0.4 sec)
```

Notice the difference when running 

`garden get status` with `deploy` instead of `dev`

```sh
garden deploy --hot-reload=website
Deploy 🚀

✔ website                   → Building website:2c12317af0... → Done (took 0.5 sec)
✔ website                   → Checking status → Version 2c12317af0 already deployed
✔ website                   → Hot reloading → Done (took 103 ms)
✔ website                   → Hot reloading → Done (took 118 ms)
✔ website                   → Hot reloading → Done (took 103 ms)
```

---

But I would expect the differences between dev and deploy + hot-reload/watch to be almost the same.

Why do we have totally different workflows/code instead of calling the same calculation logic with different params?

We should 

1. Fix the issue where their task list is calculated differently
1. Fix the documentation to highlight the differences between the `dev` and `deploy`
1. Fix issue #387

**Comment by thsig** [comment]
> 2. Fix the documentation to highlight the differences between the dev and deploy

Regarding 2. do you have any suggestions for how to change the description strings for the `dev` and `deploy` commands? At first glance, they seem pretty clear to me, but I'm perhaps a bit too familiar with how all this works. @drubin

**Comment by edvald** [comment]
@thsig Is this still an issue?

**Comment by drubin** [comment]
@thsig When I originally logged this the idea was to fix the root cause that the code is structured differently. At the time we had a bunch of issues related to this difference. tasks mismatch, hot reloading going funny, watch behaving differently. 

If you look at the commands:

* `dev` https://github.com/garden-io/garden/blob/master/garden-service/src/commands/dev.ts#L77-L137 
* `deploy` https://github.com/garden-io/garden/blob/master/garden-service/src/commands/deploy.ts#L82-L129

Their logic seems very different. Where from my understanding if you run `garden deploy -w --hotlreload... ` they should behave the same way? (maybe with tests removed)

This issue was to highlight that if we have to update this logic in 2 places every time we will always create slight differences, which was the case for a few weeks because I used `garden dev` and you used `garden deploy` 

I suspect the actual specific bug might have been resolved but the issue with the code exists in my view.
